### PR TITLE
Scan iter backport

### PR DESCRIPTION
**Pull request by wangtzh**

### CODE REVIEW

File: dbcore/sm-dir-it.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * OID Dir Iterator, used for Index query
+ */
+#pragma once
+#include "sm-config.h"
+#include "../txn.h"
+namespace ermia {
+
+struct DirIterator {
+  /* For debug use, store the oid_dir ptr */
+  OID *dirp;
```
Comment (wangtzh) [inline dbcore/sm-dir-it.h:10]: Wrap with `#ifndef NDEBUG`

File: dbcore/sm-dir-it.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        return tmpval;
+    }
+    dbtuple *tuple = nullptr;
+    tuple = oidmgr->oid_get_version(td->GetTupleArray(), o, t->xc);
+    bool found = true;
+    if (!tuple) {
+        DLOG(WARNING) << "(SKIPPED) Some tuple is empty: OID = " << std::hex << o;
+        found = false;
+    }
+    if (found) {
```
Comment (wangtzh) [inline dbcore/sm-dir-it.h]: `if (tuple)...`

File: dbcore/sm-dir-it.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        bool ret = t->DoTupleRead(tuple, &tmpval)._val;
```
Comment (wangtzh) [inline dbcore/sm-dir-it.h:39]: `DoTupleRead` returns rc_t, not boolean. If the return value indicates an abort, the calling transaction should abort. (not sure why a retry is needed here?)

File: dbcore/sm-dir-it.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        if (!ret) {
+            DLOG(WARNING) << "(SKIPPED) Cannot do tuple read for OID = " << std::hex << o;
+            idx += 1;
+            goto redo;
+        }
+	    idx += 1;
+        return tmpval;
+    } else if (config::phantom_prot) {
```
Comment (wangtzh) [inline dbcore/sm-dir-it.h]: Why is this related here?

File: dbcore/sm-oid.h [hunk 5]
```diff
@@ -317,6 +361,80 @@ struct sm_oid_mgr {
...
+    auto cur_dir = rootp;
+    while(layer != 0) {
+        cur_dir = reinterpret_cast<OID *>(oa->get(cur_dir[OID_DIR_SIZE - 1])->offset());
+        layer -= 1;
+    }
+    return cur_dir[pos];
+  }
+
+
+  // inline OID dir_get_index(oid_array *oa, OID dir_oid, uint64_t index, bool &eof) {
```
Comment (wangtzh) [inline dbcore/sm-oid.h:389]: Remove if not useful.

File: ermia.cc [hunk 6]
```diff
@@ -199,6 +201,37 @@ void ConcurrentMasstreeIndex::GetRecord(transaction *t, rc_t &rc, const varstr &
   }
 }
 
+DirIterator *ConcurrentMasstreeIndex::GetRecordMultiIt(transaction *t, rc_t &rc, const varstr &key) {
+    rc = {RC_INVALID};
+    OID dir_oid = INVALID_OID;
+    std::vector<OID> oids;
+    ermia::varstr tmpval;
+    auto dir_it = new DirIterator(t, table_descriptor);
```
Comment (wangtzh) [inline ermia.cc]: `dir_it` should be something maintained (e.g., created on stack on demand) by the caller. Frequent `new` should be avoided here.

File: dbcore/sm-dir-it.h (hunk 4 above)
Comment (VOID001) [reply dbcore/sm-dir-it.h]: Just to indicate we don't support phantom protection

File: ermia.cc (hunk 6 above)
Comment (VOID001) [reply ermia.cc]: Yes, now it don't need to be allocated every time, since the OID vector inside has gone

File: dbcore/sm-dir-it.h (hunk 4 above)
Comment (wangtzh) [reply dbcore/sm-dir-it.h]: Still doesn't look right - why is it an else branch of "found?" This branch shouldn't appear here - why can't this support phantom protection? (or why is this question relevant here?)

File: ermia.cc (hunk 6 above)
Comment (wangtzh) [reply ermia.cc]: Is this the latest up-to-date code? If not we shouldn't review it yet.

Comment (VOID001) [reply ermia.cc]: It is, I will update all changes related to the review later

Comment (wangtzh) [reply ermia.cc]: ok, but you mentioned "the OID vector inside has gone?" I assumed you had changes that aren't checked in yet.

Comment (VOID001) [reply ermia.cc]: The OID vector inside class DirIterator is gone so we don't need to allocate space for vector when creating a DirIterator, but some useless code in GetRecordMultiIt such as line 207 should also be removed.

Comment (wangtzh) [reply ermia.cc]: Still not sure what you meant: this comment was about L209, not L207.

Comment (VOID001) [reply ermia.cc]: Yes. I mean previously(before this commit) DirIterator has a OID vector inside, but now it is gone. so I can move the dir_it fully to stack now,

Comment (wangtzh) [reply ermia.cc]: ok, let's move it to stack then.

**Comment by VOID001** [comment]
Merge scan iterator support to ERMIA experimental

File: dbcore/sm-dir-it.h (hunk 3 above)
Comment (VOID001) [reply dbcore/sm-dir-it.h:39]: Sorry for the inconvenience, the changes are checked in the new branch `reopen`, basically change the ret type to rc_t and also ALWAYS_ASSERT ret is RC_TRUE
